codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/EndpointResolverMiddleware.kt

```diff
@@ -53,11 +60,32 @@ class EndpointResolverMiddleware(
         writer.write("let endpoint = try endpointResolver.resolve(params: endpointParams)")
             .write("")
 
-        writer.write("let authScheme = endpoint.authSchemes()?.first")
-        writer.write("let signingName = Endpoint.signingName(from: authScheme)")
-        writer.write("let signingRegion = Endpoint.signingRegion(from: authScheme)")
-        writer.write("let signingAlgorithm = Endpoint.signingAlgorithm(from: authScheme)")
-            .write("")
+        writer.write("var signingName: String? = nil")
+        writer.write("var signingRegion: String? = nil")
+        writer.write("var signingAlgorithm: String? = nil")
+        writer.openBlock("if let authSchemes = endpoint.authSchemes() {", "}") {
+            writer.write("let schemes = try authSchemes.map { try AuthScheme(from: \$$0) }")
+            writer.write("let authScheme = try authSchemeResolver.resolve(authSchemes: schemes)")
+            writer.write("switch authScheme {")
+            writer.write("case .sigV4(let data):")
+            writer.indent()
+            writer.write("signingName = data.signingName")
+            writer.write("signingRegion = data.signingRegion")
+            writer.write("signingAlgorithm = data.name")
+            writer.dedent()
+            writer.write("case .sigV4A(let data):")
+            writer.indent()
+            writer.write("signingName = data.signingName")
+            writer.write("signingRegion = data.signingRegionSet?.first")
+            writer.write("signingAlgorithm = data.name")
+            writer.dedent()
+            writer.write("case .none:")
+            writer.indent()
+            writer.write("break")
+            writer.dedent()
+            writer.write("}")
+        }
+        writer.write("")
```

Comment:
This is a big chunk of logic to be codegen-ed for every service.  Any way we can get this in runtime, where it can be shared among services and placed under functional test?

Reply:
can we file an issue for this? i agree we shouldn't hold up the PR for it, but it'd be good to have an issue since imo its a pretty straightforward one that anyone can pick up if they have extra time.

Reply:
it's not straightforward as we just don't want to extract the logic but want to extract the middleware outside the service client.

Indeed https://github.com/awslabs/aws-sdk-swift/issues/840